### DISCUSSION TITLE
Slow indexing

### DISCUSSION THREAD

**Discussion by AndiLeni** [loupe-php/loupe#41] [General]
I noticed, that indexing content from my CMS is extremely slow.
A regular 800 - 1000 word page takes about 3-5 seconds to index, which results in very long indexing times for a regular website with a few dozen pages.
Is there anything I can do to speed this up or find the bottlenecks?

Edit:
The bottleneck seems to be in `StateSetIndex.php` inside the function `index()`.
Calling `$this->stateSet->add($newState, $state, $mappedChar);` takes about `0.03` seconds, which accumulates to a large delay.

**Comment by Toflar** [comment]
200 documents per second is actually pretty fast 😊  There's not much to optimize here anymore, no.
So your speed is very much expected and normal. If you need faster indexing times, Loupe is not the right choice for you (it's documented in the README).

**Reply by jdreesen** [reply]
He said 3-5 sec/document, right? Not 200 docs/sec.

**Reply by Toflar** [reply]
Ah one document takes 5 seconds? That's weird then. Needs to be debugged :) But the fact that the stateset needs time is to be expected for sure.

**Reply by AndiLeni** [reply]
Yes, sorry if I made misleading statements.
One document needs about 5 seconds for indexing. 
It's just weird that it takes so long on my machine. But I was not able to identify the issue in this short time.
Just wanted to let you know that there might be issues.

**Reply by Toflar** [reply]
One beautiful thing about SQLite is that you can just zip your `loupe.db` and send it around. So in case your data is public domain and does not contain any secret information, feel free to attach it here together with the code that runs slow. I'll try to have a look when I find some time then 😊

**Reply by AndiLeni** [reply]
@Toflar I did some debugging and was digging down to where the bottleneck is.
It seems like it is the doctrine vendor.
In `Engine.php upsert()` you are calling `$this->getConnection()->executeQuery($query, $parameters)`.
This leads to the doctrine file.
Inside of `executeQuery` calling 
```php
$stmt = $connection->prepare($sql);
$this->bindParameters($stmt, $params, $types);
$result = $stmt->execute();
```
is really really slow. 
I am not sure what can be done about this or why it is slow at all.
What is the reason for loupe to use doctrine? There does not seem to be much sql which needs to be handlet on loupe's side?

**Reply by Toflar** [reply]
We do need Doctrine because of schema management. Loupe adds columns dynamically and detecting changes with the current schema makes no sense to re-implement if Doctrine does this perfectly.
Also, I don't think it's Doctrine in your case. It's the underlying PDO driver. I've experienced the same weird thing on some setup and thus decided to rely on the `sqlite3` extension if available. This is implemented in the upcoming v 0.4 (https://github.com/loupe-php/loupe/pull/43) - can you try if that makes a difference for you?

**Reply by AndiLeni** [reply]
Oh I see. I was not aware that the database scheme is dynamic.
Unfortunately, version 0.4 does not help. There must be some other issue, wich totaly might be on my side.
I did some profiling with xdebug and this is the result for inserting 1 document to the index:

```
Duration  | Function
1558.6 ms | Loupe\Loupe\Internal\Engine->upsertEngine.php:168
1466.8 ms | Doctrine\DBAL\Connection->bindParametersConnection.php:1779
1216.7 ms | Doctrine\DBAL\Connection->executeStatementConnection.php:1181
1169.9 ms | Doctrine\DBAL\Driver\PDO\Statement->bindValueStatement.php:29
1103.9 ms | Doctrine\DBAL\Connection->executeQueryConnection.php:1069
986.1 ms  | Nitotm\Eld\LanguageData->__constructLanguageData.php:42
708.3 ms  | php::PDO->prepareinternal
469.8 ms  | Doctrine\DBAL\Driver\PDO\Connection->prepareConnection.php:55
```
Do you have any ideas what else I might try?
I seems to be so slow because one field is rather large, since I try to index website text.

**Reply by AndiLeni** [reply]
@Toflar I ran some test here on my machine. The slow speed is indeed related directly to the size of the indexed document.
When I try to index a simple website article with 400 words, 2100 sql statements are executed in roughly 4 seconds.
It is not surprising that my indexing is slow. Executing multiple thousands of sql statements is really time consuming.

Maybe it is possible for you to make some adjustments here to reduce the amount of requireed sql updates? Or maybe combine them instead of executing them on their own?

**Reply by fiedsch** [reply]
>  Or maybe combine them instead of executing them on their own?

Sounds promising.  Accumulating the data data to be inserted and then executing a statement like
```
INSERT INTO table (column1,column2 ,..)
VALUES 
   (value1,value2 ,...),
   (value1,value2 ,...),
    ...
   (value1,value2 ,...);
```
might help to reduce the number of queries and speed up things a bit.

**Reply by fiedsch** [reply]
An alternative might be to use `INSERT` together with `ON CONFLICT` (https://www.sqlite.org/lang_conflict.html) to reduce the overhead of testing wether to insert or update the data in `Engine::upsert()`. Disclaimer: I‘m afraid I don not understand the data structures enough to see if this would make sense :-(

**Reply by Toflar** [reply]
It's not like I haven't tried all of your proposals already - even before I've open sourced Loupe 😉  `upsert()` is used because the resulting ID is required for relations in those places. So using bulk inserts is not an option there. And the `ON CONFLICT` is something I've tried but failed - it's [even mentioned in the source code](https://github.com/loupe-php/loupe/blob/2fc7306899d89ed8e71a21c9636b9871d7951699/src/Internal/Engine.php#L161-L162) so if anybody is successful - PRs welcome. I'm sure there is potental to group stuff, take adavantage of built-in SQLite features, foreign keys or what not but it may require significant time investments here.

But then again: 2100 SQL statements for 400 words is not surprising at all. The data is split and optimized for searching, that's why indexing is slow **and it's not a bad thing** - it's supposed to be slow. Indexing will always be slow because that's the nature of how indexing works. Not saying we cannot optimize it but you have to free yourselves from thinking that we can combine this into one simple query. That's just not how things work.

Also, we have a performance test in the repo: If I run `bin/index_performance_test.php`, it finishes in 230s and in doing so, it indexes 32k documents, 152k terms, 2.7 million relation rows and more! So that's a hell lot of data SQLite processes within 230s. And not to mention the whole state set calculation and language detection which takes up a significant share of those 230s for sure.

So 400 words taking 4 seconds sounds a bit strange - maybe you can create a reproducer for us - should look about the same as the `bin/index_performance_test.php` with your data and then we can measure if it takes 4 seconds on our systems as well?

**Reply by AndiLeni** [reply]
@Toflar I was able to identify the issue. The slow indexing time was caused by the xdebug php extension. Even though it was not actively profiling the indexing process, it did slow down the process drastically. After disabling the extension in the php.ini file, the indexing was way faster (about 1 second for one website page).
